### ISSUE TITLE
Problem of combination dependsOnMethods and RetryAnalyzer

### ISSUE THREAD

**Issue by qwikso** [cbeust/testng#1538]
Hi guys,

when a test failed and passed after retry using IRetryAnalyzer implementation, all dependent tests are skipped. This seems to be a bug. Have a look at my simple example below.

### TestNG Version
latest (6.12)

### Expected behavior
1. a() failed //i==0
2. a() passed //i==1
3. b() passed //i==2

### Actual behavior
1. a() failed //i==0
2. a() passed //i==1
3. **b() skipped** // this is the problem

### Is the issue reproductible on runner?
yes
- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
```java
public class Example {
    int i = 0;

    public static class Retry implements IRetryAnalyzer {
        @Override
        public boolean retry(ITestResult result) {
            return true;
        }
    }

    @Test(retryAnalyzer = Retry.class)
    public void a() {
        Assert.assertTrue(i++ == 1);
    }

    @Test(dependsOnMethods = "a", retryAnalyzer = Retry.class)
    public void b() {
        Assert.assertTrue(i++ == 2);
    }
}
```

**Comment by ps231** [comment]
The ITestResult in RetryAnalyzer, keeps the test result for the failed test, thus, skipping the dependent tests. Removing that test result in the RetryAnalyzer helped me run the dependent tests, if the original test passed on a retry.

My simple implementation of the RetryAnalyzer.retry() API - 

`@Override
    public boolean retry(final ITestResult result) {
        if (current <= maxRetryCount) {
            result.getTestContext().getFailedTests().removeResult(result);
            current++;
            LOGGER.info("Retrying test case " + result.getMethod() + ", " + current + " times");
            return true;
        }
        LOGGER.info("Exhausted retries for test case " + result.getMethod());
        return false;
    }`

FYI, I am using 6.8.7 version of TestNG.

**Comment by strifu** [comment]
@jigyasa-at-git Any update to this? I am having the exact issue.

Thanks!

**Comment by strifu** [comment]
@cbeust Can you provide any insight into the issue being discussed here?

**Comment by gowthami7** [comment]
Any update on this issue?

**Comment by strifu** [comment]
@gowthami7  Yes, added:
result.getTestContext().getFailedTests().removeResult(result);
result.getTestContext().getSkippedTests().removeResult(result);

To an  @AfterMethod instead and it worked like a charm.

**Comment by gowthami7** [comment]
Thank you. It worked!

**Comment by JasonBlevins29** [comment]
Where does the @aftermethod go? Does it go in the IRetryAnalyzer class or in the test class?

**Comment by krmahadevan** [comment]
@gowthami7 @JasonBlevins29 @strifu @ps231 @jigyasa-at-git @qwikso - This now stands fixed in the upcoming TestNG version (7.0.0). TestNG will now automatically execute dependent methods, once their parent methods have been retried and run succeeded.  You can try depending on `TestNG 7.0.0-SNAPSHOT` and try this out now. (Make sure you are using JDK8 because from 7.0.0 TestNG now requires JDK8)

**Comment by neha1302** [comment]
@gowthami7 @strifu 
I tried using both the options suggested by you. Adding @AfterMethod and using the IInvokedMethodListener  Listener as well. I am still facing the same issue. Retry is not working for the dependent test cases. 
Could you please suggest.

**Comment by krmahadevan** [comment]
@neha1302 - This was a bug that needed code fix and it stands resolved in `7.0.0-SNAPSHOT`. Are you still facing this problem with the latest version ? If not, then you just need to wait till it gets released.

**Comment by neha1302** [comment]
@krmahadevan 
Thanks for answering!
As far as I know 7.0.0 is not released yet. I am using testNG-6.14.3

**Comment by krmahadevan** [comment]
@neha1302 - Yes I am very much aware that its not released yet. That was why I was asking, if you tried using TestNG `7.0.0-SNAPSHOT` ? Can you please try using that and let us know if using it fixes your problem ? If it doesn't, you would need to file a new issue with more details so that we can go about fixing your issue.

**Comment by neha1302** [comment]
@krmahadevan 
Ok. 
Actually I do not know what exactly 7.0.0-SNAPSHOT is? and how shall I get it?
Couldn't find much on Google.
Please suggest.

**Comment by krmahadevan** [comment]
@neha1302 
If you are using Maven, you should just need to add a dependency which looks like below

```xml
        <dependency>
            <groupId>org.testng</groupId>
            <artifactId>testng</artifactId>
            <version>7.0.0-SNAPSHOT</version>
        </dependency>
```

**Comment by krmahadevan** [comment]
@neha1302 - Are you behind any corporate artifactory repository ?  If yes then you may have to google for how to create a `settings.xml` equivalent in gradle. [This](https://discuss.gradle.org/t/gradle-settings-xml-file-equivalent-from-maven-to-define-a-global-local-repository/11158/2) may help. I have not worked with gradle elaborately to be able to tell you as to how to make it point to the Maven snapshots repository

Here's a sample settings.xml which refers to the maven snapshots https://gist.github.com/ktuukkan/8cf2de1e915185118c60 You still need to tweak it to change the active profile value to snapshots profile name, or pass the profile name explicitly (using gradle equivalents of course)

**Comment by neha1302** [comment]
@krmahadevan 
Ok. Will do that. Meanwhile Can I just download it from somewhere and place it in my project, just to see whether I am able to execute my dependent test cases with this or not?
If yes, please tell me from where to download.

**Comment by krmahadevan** [comment]
Here's where its available https://oss.sonatype.org/content/repositories/snapshots/org/testng/testng/7.0.0-SNAPSHOT/

**Comment by neha1302** [comment]
@krmahadevan 
I added jar - testng-7.0.0-20180713.040141-27.jar
I tried executing the test cases using both ways mentioned in above discussion i.e.
1. Adding the code in the @AfterMethod
2. Using IInvokedMethodListener

Still if I am failing in TC2 which is dependent on TC1, the TC1 is not executing. The retry again starts from TC2.
I have added the IInvokedMethodListener implementing class in the xml file as well.

Let me know if I am missing something or there is some gap in my understanding.

**Comment by krmahadevan** [comment]
@neha1302 - Couple of things.. 

The latest jar is 
https://oss.sonatype.org/content/repositories/snapshots/org/testng/testng/7.0.0-SNAPSHOT/testng-7.0.0-20180725.174808-30.jar

How are you ensuring that your project is indeed using TestNG 7.0.0-SNAPSHOT ? 

If you are convinced that your problem exists, please file a new bug and include a simple standalone that can be used to reproduce the problem.

**Comment by neha1302** [comment]
![testng](https://user-images.githubusercontent.com/23551859/43627041-399e7ede-9712-11e8-8014-14111fd3ba07.png)
This is how I am ensuring about the jar.

**Comment by krmahadevan** [comment]
@neha1302 - And have you cleaned up your `build.gradle` to remove any references of TestNG ? Also have you made sure that TestNG eclipse plugin doesn't use the jar that is embedded inside it. Earlier there used to be a way to force it not to use the embedded jar. But I dont know if it still exists (I stopped using eclipse a long time back). That is why I feel that the easiest and surest way would be add it as a dependency via your gradle and then try running tests from gradle command prompt.

**Comment by neha1302** [comment]
![testngjar](https://user-images.githubusercontent.com/23551859/43628059-72b215f6-9716-11e8-920a-5e308240f03b.png)
I am getting this one as well.
Do you mean to say this could be the culprit?

**Comment by krmahadevan** [comment]
@neha1302 - Yes very much. You need to absolutely ensure that you are **ONLY** resolving to TestNG `7.0.0-SNAPSHOT`

**Comment by neha1302** [comment]
I am unable to remove this. How shall I remove this?

**Comment by krmahadevan** [comment]
You need to figure out what dependency is bringing in an old version of TestNG and add an exclusion to it as discussed [here](https://discuss.gradle.org/t/how-do-i-exclude-specific-transitive-dependencies-of-something-i-depend-on/17991)

**Comment by neha1302** [comment]
![jni error](https://user-images.githubusercontent.com/23551859/43641712-550d7b04-9742-11e8-94e1-27a5dd37f4e6.png)
Hi,
I have removed the previous testng dependency and added the required jar.
I am getting this error now and am not able to execute test cases.

**Comment by krmahadevan** [comment]
Neha,

I think its better if you just update your dependencies with TestNG
7.0.0-SNAPSHOT. That will be easier for you. So please search for how to
add a SNAPSHOT dependency and follow those instructions to update your
gradle build file with it.

The problem here is that you downloaded TestNG jar and added it to your
CLASSPATH (TestNG internally needs jcommander but since you added an
exclusion to TestNG, even JCommander got excluded). So now the TestNG jar
in your CLASSPATH needs JCommander and its not able to find it, that is why
you are seeing this error.

**Comment by neha1302** [comment]
![jcoomandererror](https://user-images.githubusercontent.com/23551859/43706570-631b430a-9983-11e8-8b8e-d9cdd9ef43ef.png)
@krmahadevan 
I am still getting the same error.

**Comment by krmahadevan** [comment]
In that case, please create a new issue on github and include a simple (its
better if it just uses a bunch of system.out.println statements) standalone
sample which can be executed to recreate the problem, along with what is
being expected and what are you seeing actually.

>

**Comment by qwikso** [comment]
Hi, I've tested version 7.0.0-SNAPSHOT a month ago and from my point of view, everything works fine. The example from the description at the top works as expected. If you copy the snippet to your environment do you get still the unexpected result?

We are using Maven instead of Gradle and there is a configuration for local jars. I downloaded snapshot version and added this config to my pom.xml. I believe Gradle has a similar solution.

`
<dependency>
    <groupId>org.testng</groupId>
    <artifactId>testng</artifactId>
    <version>7.0.0</version>
    <scope>system</scope>
    <systemPath>${basedir}/lib/testng-7.0.0.jar</systemPath>
</dependency>
`

**Comment by neha1302** [comment]
@qwikso 
I am pulling 7.0.0 from the gradle dependencies to check the rerun. The problem now is I am not able to execute any of the TestNG XML file because I am getting jcommander issue mentioned in the above posts because of which I am not able to work on rerun.
Did u also got the same issue while working on this?

**Comment by neha1302** [comment]
@krmahadevan 
That jCoomander issue got resolved but I am getting the below error:
java.lang.NoSuchMethodError: org.testng.remote.AbstractRemoteTestNG.addListener(Lorg/testng/ISuiteListener;)V
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)

Is this because 7.0.0 is not supported by Eclipse?
What should be my next step to deal with it?

**Comment by krmahadevan** [comment]
The root cause of your problem is STILL the same. A messed up classpath.
Just to rule out all confusions, I would strongly advice you to consume
TestNG 7.0.0-SNAPSHOT as a gradle dependency and try running the tests from
the command line using gradle.

That is going to be a lot more easier. Have you tried doing that ? If not,
I strongly encourage you to do it.

**Comment by neha1302** [comment]
I am taking TestNG as a gradle dependency only and am trying to execute the test cases through Eclipse.

**Comment by krmahadevan** [comment]
Can you please skip using an IDE and just run tests from the command prompt
using Gradle ?

**Comment by qwikso** [comment]
No such problem in my environment. Running from IntelliJ as well as from Maven command line works as expected. I would create a greenfield project with minimum dependencies and run it using Gradle from command line without IDE as @krmahadevan does recommend.

**Comment by neha1302** [comment]
@krmahadevan 
My scenario is:

 @Test
testCase1()
{
}
 @Test(dependsOnMethods = { "testcase1" })
testCase2()
{
}

 @Test(dependsOnMethods = { "testCase2" })
testCase3()
{
}


If testCase2 fails, testCase1 should be executed.
Is it possible?

**Comment by krmahadevan** [comment]
No. That is not possible. The fact that *'testCase2"* got executed the
first time itself proves that its pre-requisites [*testCase1* should pass]
were satisfied.
So TestNG will not bother trying to satisfy the pre-requisites once again,
just because *testCase2* failed and needs to be retried.
